### PR TITLE
[ROCm] aiter mha kernels (ASM+CK) integration (#747)

### PR DESCRIPTION
**Pull request by zahiqbal**
## Summary

This PR adds AMD AITER's flash-attention kernels (both Composable-Kernel and ASM v3 backends) as a first-class JAX primitive on ROCm. The kernels are exposed through the FFI as `hip_mha_fwd_ffi` / `hip_mha_bwd_ffi` and wrapped on the Python side by `flash_attn_func` (batch) and `flash_attn_varlen` (variable-length), both with `custom_vjp` for autograd and `custom_partitioning`.
The pre-built AITER shared libraries and headers are fetched as a Bazel external repo (`@aiter_mha_wheel`) from the `ROCm/jax` GitHub release `AITER-MHA`, or sourced from a local checkout via `LOCAL_AITER_PATH` for development.

## Motivation
JAX on ROCm currently has no production-grade fused attention kernel. AITER provides tuned CK and ASM v3 kernels for gfx942 / gfx950. Integrating them directly into `jaxlib` (rather than as an out-of-tree plugin) means:
- one wheel ships everything users need (`libmha_fwd.so`, `libmha_bwd.so` are bundled into the `jax-rocm-*-plugin` wheel),
- FSDP / batch-sharded attention works through `custom_partitioning` without any user-side glue,
- Python users get a stable `jax._src.aiter` API that does not depend on PyTorch / `aiter` Python package.

## Public API
```python
from jax._src.aiter import flash_attn_func, flash_attn_varlen
```

Both functions:

support fp16 / bf16, head dims 32–256, MHA / MQA / GQA,
accept causal, sliding-window (window_size_left, window_size_right), softmax_scale, dropout_p, ALiBi slopes, attention bias, logits soft-cap,
expose use_asm_v3 to opt into the ASM v3 backend (with how_v3_bf16_cvt, is_v3_atomic_fp32 knobs),
return either the output alone or (out, lse, ...) when return_lse / return_randval are set,
are differentiable through custom_vjp, and partitionable through custom_partitioning (output sharding follows Q, no collectives — every dim except batch is replicated per device).
If the AITER .so are not present at runtime (e.g. CUDA build, or stripped wheel), the public symbols import as None so the rest of jax still loads.

## What's in the PR

FFI handlers (C++):

- jaxlib/gpu/aiter_mha_fwd.cc — unified forward handler (aiter::mha_fwd), batch + varlen detected from rank.
- jaxlib/gpu/aiter_mha_bwd.cc — unified backward handler (aiter::mha_bwd).
- jaxlib/gpu/aiter_mha_common_utils.{cc,h} — dtype / stride / mask / bias helpers shared between fwd and bwd.
- jaxlib/gpu/aiter.{cc,h} — nanobind module _aiter exporting hip_mha_fwd_ffi / hip_mha_bwd_ffi registrations.

Python:

- jax/_src/aiter/aiter_mha.py — flash_attn_func, flash_attn_varlen, custom_vjp, custom_partitioning, FFI wrappers, GFX detection.
- jax/_src/aiter/__init__.py — public re-exports with graceful fallback when the kernels are unavailable.
- jaxlib/gpu_aiter.py — plugin-aware loader that pulls _aiter out of the installed ROCm plugin wheel.
- jax/_src/lib/__init__.py — best-effort import jaxlib.gpu_aiter (falls back to None).

Build / packaging:

- WORKSPACE, third_party/aiter/workspace.bzl, third_party/aiter/BUILD — fetch the AITER MHA wheel by URL+SHA256, expose libmha_{fwd,bwd}.so and headers, support LOCAL_AITER_PATH override (symlinks a local aiter_mha-*.data directory).
- jaxlib/rocm/BUILD — new aiter_mha rocm_library and _aiter nanobind extension; added to rocm_gpu_support.
- jaxlib/gpu/BUILD, jaxlib/BUILD, jax/BUILD, jax/_src/BUILD — wire the new sources / py targets.
- jaxlib/tools/build_gpu_kernels_wheel.py — copy libmha_{fwd,bwd}.so into the ROCm plugin wheel and fail the build if either is missing, so a broken wheel can never ship.
- jaxlib/tools/build_wheel.py, jaxlib/tools/BUILD.bazel — include gpu_aiter.py in jaxlib and the AITER .sos in wheel_sources.
- build/build.py, build/rocm/dev_build_rocm.py, build/rocm/tools/build_wheels.py — new --local_aiter_path / --local-aiter-path CLI flag that forwards to --repo_env=LOCAL_AITER_PATH=....

Tests:

- tests/aiter_mha_test.py (570+ lines) — gated on ROCm devices, covers:

- - all supported head dims (32, 40, 59, 64, 96, 111, 128, 160, 256),
- - MHA / MQA / GQA, self-attention, cross-attention, off-by-one shapes, decode (sq=1), batch=8,
- - varlen (cu_seqlens) variants of the above,
- - causal, SWA, dropout (shape/finiteness only),

### CODE REVIEW

File: jaxlib/gpu/aiter_mha_common_utils.cc [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+          batch_size * num_heads * ck_tile::get_warp_size());
+
+      VLOG(1) << "Generated RNG with seed: " << seed_value
+              << ", offset: " << offset_value << " (no gen provided)";
+
+      uint64_t host_rng[2] = {seed_value, offset_value};
+      hipError_t err =
+          hipMemcpyAsync(rng_base, host_rng, 2 * sizeof(uint64_t),
+                         hipMemcpyHostToDevice, stream);
+
```
[the remark this answers is not in the document]
Comment (zahiqbal) [reply jaxlib/gpu/aiter_mha_common_utils.cc]: hipMemcpyAsync is changed to hipMemcpy

File: jaxlib/gpu/aiter_mha_common_utils.h [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+} // namespace jax_aiter
+
+namespace jax_aiter {
+
+  inline void hipCheck(hipError_t err, const char *file, int line) {
+    if (err != hipSuccess) {
+      LOG(ERROR) << "HIP error at " << file <<":"<< line << ": " << hipGetErrorString(err);
+      std::abort();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (zahiqbal) [reply jaxlib/gpu/aiter_mha_common_utils.h:263]: throw std::runtime_error is used instead of abort.

File: jax/_src/aiter/aiter_mha.py [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+
+if gpu_aiter is not None:
+    for platform, targets in gpu_aiter.registrations().items():
+      for name, value in targets:
+        print(f"Registering AITER FFI target: {name} for platform: {platform}")
+        jax.ffi.register_ffi_target(
+            name, value, platform=platform, api_version=1
+        )
+else:
+    print("AITER FFI targets not found")
```
[the remark this answers is not in the document]
Comment (zahiqbal) [reply jax/_src/aiter/aiter_mha.py]: using logger instead of print

File: jaxlib/gpu/aiter_mha_common_utils.cc [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  if (tid >= total_elements)
+    return;
+
+  int b = tid / (seqlen * num_heads_k * head_dim);
+  int remainder = tid % (seqlen * num_heads_k * head_dim);
+  int s = remainder / (num_heads_k * head_dim);
+  remainder = remainder % (num_heads_k * head_dim);
+  int hk = remainder / head_dim;
+  int d = remainder % head_dim;
```
[the remark this answers is not in the document]
Comment (zahiqbal) [reply jaxlib/gpu/aiter_mha_common_utils.cc]: Done, int64_t is used..
